Basic NFC Modem functionality for Pocket Computer

hugeblank opened the pull request:
Attaches a 0-block radius "NFC" wireless modem to the top of the pocket computer. This is done to address the core issue presented in both #1406 and #1688.

TODO:
- [ ] ~~Rename peripheral type to "nfc"~~
- [x] Use per-player wireless network instead of the global one to prevent ender modem sniffing of NFC transmissions.
- [ ] ~~Rename transmission events to "nfc_message"~~

Feedback would be greatly appreciated on this to do list, specifically whether any/all of the items should actually be done.

hugeblank committed:
Basic NFC Modem functionality

Wojbie commented:
#### Feedback

>  Rename peripheral type to "nfc"
>  Rename transmission events to "nfc_message"

Does it really need to implement as separate logic? I mean if this was just a normal wireless modem it would work with all already written code (and rednet).

> Use per-player wireless network instead of the global one to prevent ender modem sniffing of NFC transmissions.

We could achieve same effect just by not allowing messages from this modem to be heard on range over 1 block. Not sure how feasible that would be to invert this bit of message networking logic but it might be less complicated then creating completely new system.

#### Suggestion

I would suggest increasing range to 1 block to allow near range communication
This way while adding ability to network inside player inventory we also add ability to communicate between 2 players close to each other (like in a boat) [Sneakernet anyone?] or between player and wireless modem right next to it (commonly used for door locks in CC:T).

Wojbie commented on projects/common/src/main/java/dan200/computercraft/shared/pocket/peripherals/PocketNFCPeripheral.java:
```diff
@@ -0,0 +1,30 @@
+        setLocation(access);
+
+        var state = getModemState();
+        if (state.pollChanged()) access.setLight(state.isOpen() ? 0xf39d20 : -1);
```
Traditionally only main peripherals set light status. This might look odd if this one and back peripheral will both try to set different colours.

hugeblank replied:
Good point, not sure how to address it though. Color blending? Blinking between the two colors?

hugeblank replied:
5c67b6c is my attempt at addressing this, the light slowly cycles between the two active colors

fatboychummy commented:
> I would suggest increasing range to 1 block to allow near range communication

Personally I'd say 2 blocks. Depending on where the range is checked from, you may end up in situations like the following: (assuming it checks from the feet)

![SmartSelect_20240119_152448_Samsung Notes](https://github.com/cc-tweaked/CC-Tweaked/assets/39929480/8b6e6096-948f-4a43-96f8-1f8b5903eea3)

Where gray is range of 1, purple is range of 2. If it checks from the feet, a computer at head-height would be out of range with a range of 1, even if you stood right beside it.

Wojbie commented:
> IMO, the objective here is strictly to have a means to transmit messages contained within the player's inventory. If it's desired to have out-of-inventory communication, then that's when you'd attach a wireless modem.

I see your point. And in the end if someone really wanted long range modem communication they could just carry a wireless pocket that would repeat messages between wireless modem and this inner modem. I just think that combining this with short range (1-block) mode from #1688 would provide more wide game play possibilities. Like sneakernet or pokemon style pocket gaming.

> Personally I'd say 2 blocks. Depending on where the range is checked from, you may end up in situations like the following: (assuming it checks from the feet)

If i remember right CC modems transmit from eyes of player.

hugeblank commented:
> I see your point. And in the end if someone really wanted long range modem communication they could just carry a wireless pocket that would repeat messages between wireless modem and this inner modem. I just think that combining this with short range (1-block) mode from #1688 would provide more wide game play possibilities. Like sneakernet or pokemon style pocket gaming.

I agree, I just don't think it's in the scope of this PR, and more-so something that should be added onto wireless modems like proposed in the referenced issue.

hugeblank committed:
NFC Network Isolation

tested as functional in multiplayer

hugeblank commented:
Latest commit isolates networks on a per-entity basis, preventing sniffing from ender modems.

hugeblank committed:
Attempt to pass tests

hugeblank committed:
Pocket computer light color cycle

Slowly cycles between the orange NFC modem color, and the color set by the back peripheral if both are active.

hugeblank committed:
Fix light being too dim

hugeblank commented:
Demonstration of the light cycle effect:
https://github.com/cc-tweaked/CC-Tweaked/assets/16393543/298d10df-0044-40f2-aedb-331b0402f9e3

This is a bit extreme, but I thought it would be interesting to see what it looks like and field discussion from there.

hugeblank committed:
Fix style
